[TypeChecker] SR-7307 : Warn when a block contains defer as last statement

vguerra opened the pull request:
This PR adds a diagnostic (~~NOTE~~ WARNING ) when a defer statement is the last statement within a block of code. The diagnostic is implemented as follows:

- During type-checking a `BraceStmt` simply check if last statement of it's elements is a `DeferStmt`. If that is the case, emit diagnostic.
- A flag (`IsBraceStmtFromTopLevelDecl`) has been added to the `StmtChecker` in order to keep track of whether the `BraceStmt` being checked belongs to a `TopLevelCodeDecl` or not. This is needed to avoid diagnosing every `defer` statement found at the top level. (check discussion in : [SR-7307](https://bugs.swift.org/browse/SR-7307))
- As a consequence of prev. point, this solution is not able to diagnose a `defer` statement that appears last at the top level ( not sure if this is worth adding ).

Resolves [SR-7307](https://bugs.swift.org/browse/SR-7307).

vguerra committed:
[TypeChecker] SR-7307 : Diagnose defer statments that appear at the end block.

xwu commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -54,6 +54,9 @@ NOTE(previous_decldef,none,
 NOTE(brace_stmt_suggest_do,none,
      "did you mean to use a 'do' statement?", ())
 
+NOTE(defer_stmt_at_blockend,none,
```
Nit: "block_end" (two words).

vguerra replied:
Fixed in next commit

xwu commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -54,6 +54,9 @@ NOTE(previous_decldef,none,
      "did you mean to use a 'do' statement?", ())
 
+NOTE(defer_stmt_at_blockend,none,
+     "defer statement at the end of block is pointless", ())
```
Pointless because...? It's important to give a little more explanation here, I think. Also, stylistically, "defer" should be surrounded by single quotes, as shown above with "a 'do' statement".

jszumski replied:
I'd suggest something like:
> 'defer' at the end of its scope is redundant and will execute immediately

Probably best for a follow up PR, but it would be nice to have a fixit that simply removes the `defer` and braces.

xwu replied:
It seems it'd be simpler just to suggest `do` in place of `defer`.

vguerra replied:
thanks @xwu and @jszumski for the reviews. I improved the note message as you suggested in the following commit.

I thought as well about the fix-it from the beginning but did not have the time to see how to implement it, I am still getting familiar with the code base :). As @xwu, probably the easier/simpler thing would be to suggest a `do`. Will look into it.

vguerra replied:
hi @xwu, @jszumski 
I pushed a 3rd commit that simply adds a fix-it to the diagnostic. The fix-it just suggest to replace the `defer` with a `do` statement.

xwu commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -352,13 +355,13 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
 
   StmtChecker(TypeChecker &TC, AbstractFunctionDecl *AFD)
-    : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false) { }
+    : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false), IsBraceStmtFromTopLevelDecl(false) { }
```
Formatting nit: please wrap to 80 characters per Clang style.

vguerra replied:
Fixed in next commit.

xwu commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1333,6 +1336,19 @@ void TypeChecker::checkIgnoredExpr(Expr *E) {
   const SourceManager &SM = TC.Context.SourceMgr;
+
+  // Diagnose defer statement being last one in block ( Only if
+  // BraceStmt does not start a TopLevelDecl )
```
Formatting nit: please end comments with a period here and elsewhere; also, no spaces or capitalization inside parentheses.

vguerra replied:
Fixed in next commit.

xwu reviewed (commented):
Mostly some formatting nits. My two substantive pieces of feedback are:

The diagnostic should, I think, be more descriptive; a user who has the misunderstanding described in the linked bug won't be educated by seeing a diagnostic that says "pointless." Maybe something along the lines of saying that the body will not be deferred because it's already at the end of the block.

My second concern is how this interacts with initializers. As you know, setting a property inside an initializer doesn't invoke any didSet or willSet routines. However, whether a bug or no, I recall there's been some issue where a defer statement can be used inside an initializer to get around that. I don't recall if that's still the case. But, if so, a defer statement at the end of an initializer isn't actually pointless as removing it would change the observed behavior.

vguerra committed:
Fixing formating and a more clear explanation for diagnostic.

CodaFi commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -352,13 +355,13 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
 
   StmtChecker(TypeChecker &TC, DeclContext *DC)
-    : TC(TC), TheFunc(), DC(DC), IsREPL(false) {
+    : TC(TC), TheFunc(), DC(DC), IsREPL(false), IsBraceStmtFromTopLevelDecl(true) {
```
This is not necessarily `true`.  You'll want to ask `isa<TopLevelCodeDecl>(DC)`.

vguerra committed:
Adding a fix-it to diagnosting to replace 'defer' with a 'do' statement.

rudkx commented:
I'm not aware of other cases where we emit a note in isolation as opposed to being attached to a warning or error.

Do you know if there is precedence for this?

rudkx reviewed (changes requested):
Please add a test that the fix-it is applied correctly.

CodaFi commented:
> I'm not aware of other cases where we emit a note in isolation

Oh, good catch.  No, that is never correct - and I thought the diagnostic engine would trap on that.  This should be a warning.

CodaFi commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -54,6 +54,10 @@ NOTE(previous_decldef,none,
 NOTE(brace_stmt_suggest_do,none,
      "did you mean to use a 'do' statement?", ())
 
+NOTE(defer_stmt_at_block_end,none,
```
This should be a warning.  Notes are always attached to warnings or errors.

vguerra replied:
This is fixed in the next commit :)

vguerra committed:
NOTE should actually be a WARNING.

Updating test as well to check that the fix-it is applied correctly.

vguerra commented:
hi @rudkx! .. thank you for catching this. In my last commit I converted the NOTE into a WARNING and updated the test file to check for the fix-it!

rudkx commented:
@swift-ci Please smoke test

vguerra committed:
Fixing broken tests.

Usage of 'defer' now emits a warning in some places in the tests,
so adding the corresponding 'expected-warning'.

rudkx commented:
@swift-ci Please smoke test

vguerra commented:
a test that has been added recently fails this time :( ( https://ci.swift.org/job/swift-PR-Linux-smoke-test/8474/testReport/junit/Swift(linux-x86_64)/stmt/yield_swift/ ). My bad, I did not rebase locally before running tests locally.

fix is coming :)

xwu commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -54,6 +54,10 @@ NOTE(previous_decldef,none,
      "did you mean to use a 'do' statement?", ())
 
+WARNING(defer_stmt_at_block_end,none,
+     "'defer' at the end of its scope is redundant and will execute immediately; "
```
@jrose-apple wrote an [excellent guide](https://github.com/apple/swift/blob/master/docs/Diagnostics.md) on Swift diagnostic messages. Some points to consider here include using a terse style (known as reduced written register), phrasing as a rule, and matching the tone of other diagnostics. So perhaps:

> 'defer' statement before end of scope always executes immediately; replace with 'do' statement to silence this warning

vguerra replied:
thank you for the pointer to the Diagnostics guide... useful stuff :+1: .. and yes the message you suggest definitely reads better. Will update it in next commit.

vguerra replied:
warning message improved in ffd13662d11fb1930fe4336cb8aef90d66175a35

hamishknight commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -54,6 +54,10 @@ NOTE(previous_decldef,none,
      "did you mean to use a 'do' statement?", ())
 
+WARNING(defer_stmt_at_block_end,none,
+     "'defer' at the end of its scope is redundant and will execute immediately; "
```
Nit: The diagnostic definition is better placed within DiagnosticsSema.def as it's only emitted from Sema. DiagnosticsCommon is for common diagnostics that are emitted from separate parts of the compiler (e.g Parse and Sema).

rudkx replied:
Ah, good catch. Yes, this belongs in `DiagnosticsSema.def`.

vguerra replied:
thank you for catching this ... fixed in ffd13662d11fb1930fe4336cb8aef90d66175a35

vguerra committed:
addressing comments on PR

- moving WARNING definition from DiagnosticsCommon.def to DiagnosticsSema.def
- improving WARNING message to better align with guidelines described in docs/Diagnostics.md as suggested by @xwu.
- adapting expected warning message in tests.

rudkx commented on test/stmt/yield.swift:
```diff
@@ -82,7 +82,7 @@ func call_yield() {
   var property: String {
     _read {
-      defer {
+      defer { // expectd-warning {{'defer' statement before end of scope always executes immediately}}{{7-12=do}}
```
typo: expectd

vguerra replied:
😢 ... fixed on ab513c8abb7eb20e89ea8f2bb50a409788f843ff

vguerra committed:
typo on expected-warning.

rudkx commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -352,13 +355,16 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
   StmtChecker(TypeChecker &TC, DeclContext *DC)
-    : TC(TC), TheFunc(), DC(DC), IsREPL(false) {
+      : TC(TC), TheFunc(), DC(DC), IsREPL(false),
+        IsBraceStmtFromTopLevelDecl(true) {
```
I missed this earlier. Rather than setting this to `true`, can you instead check `getContextKind()` of the `DeclContext` that is passed in. That is less fragile, since as written this relies on the fact that the only place we currently use this constructor is when we're passing in a top level decl context.

vguerra replied:
I think this is kind of similar to what Robert suggested above about using `isa<TopLevelCodeDecl>(DC)`,  to what I responded:

> Unfortunately I can not relay on `isa<TopLevelCodeDecl>(DC)` because in the following code snippet, it would return `true` when visiting the `{` that opens the `if`s `then` block:

```swift
let x = 1
let y = 3

if (x > y) {
    defer {
        print("not so useful defer stmt.")
    }
}
```
> and I would like to diagnose this block.

when visiting the open brace of the if `DC->getContextKind()` returns `TopLevelCodeDecl`.

rudkx replied:
For that you should have an outer `BraceStmt`, with an `IfStmt` nested within it, and a `BraceStmt` nested under it. So I would expect that when you hit the outermost `BraceStmt` you would set `IsBraceStmtFromTopLevelDecl` to false like you are now.

vguerra replied:
sorry.. I completely missed your point. I see what you mean now, I should base the initialization of the IsBraceStmtFromTopLevelDecl on `DC->getContextKind()`. Fixed it on da3149d713f5c9e7b1faed096482c66e7c70d0e4

vguerra committed:
taking into account DeclContext's kind for initialization.

If DeclContext's kind is TopLevelCodeDecl we are sure to be visiting
a BraceStmt that opens a TopLevelDecl.

rudkx commented on test/Sema/diag_defer_block_end.swift:
```diff
@@ -0,0 +1,23 @@
+// RUN: %target-swift-frontend -typecheck -verify %s
```
You can take care of this in a future commit, but this could actually be `// RUN: %target-typecheck-verify-swift` without explicit options or `%s`.

rudkx reviewed (approved):
LGTM.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@vguerra I opened https://github.com/apple/swift/pull/18277 to fix up a bunch of run lines that could be using `%target-typecheck-verify-swift`.

vguerra commented:
thanks @rudkx !
